Parse symbolic shapes

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69871
* #71170
* #71121
* #69776
* __->__ #69775
* #69645

Adds parsing for Symbolic Shapes.

Differential Revision: [D33515233](https://our.internmc.facebook.com/intern/diff/D33515233)

Commit:
Parse symbolic shapes

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/frontend/schema_type_parser.cpp:
```diff
@@ -206,7 +206,7 @@ TypePtr SchemaTypeParser::parseRefinedTensor() {
   parseList(TK_NOTHING, ',', ')', [&] {
     // Extra handling for options like 'device' and 'requires_grad'
-    if (L.cur().kind == TK_IDENT) {
+    if (L.cur().kind == TK_IDENT && L.cur().text() != "SS") {
```
Nit: There is a comment in the beginning showing what parser supports, might be worth mentioning symbolic shapes in examples as well.

ZolotukhinM commented on torch/csrc/jit/frontend/schema_type_parser.cpp:
```diff
@@ -257,10 +257,21 @@ TypePtr SchemaTypeParser::parseRefinedTensor() {
     // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
     std::string::size_type num_len;
-    auto dim = c10::stoll(num, &num_len);
+    int64_t dim = c10::stoll(num, &num_len);
```
Just to double check: when we pass `[10, -2, 30]` as `dims`, `TensorType::create` will figure out that it needs to create a symbolic shape for -2, right?

eellison replied:
yup

ZolotukhinM reviewed (approved):
Looks good!

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33515233).

Commit:
Update on "Parse symbolic shapes"


Adds parsing for Symbolic Shapes.

Differential Revision: [D33515233](https://our.internmc.facebook.com/intern/diff/D33515233)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33515233).

Commit:
Update on "Parse symbolic shapes"


Adds parsing for Symbolic Shapes.

Differential Revision: [D33515233](https://our.internmc.facebook.com/intern/diff/D33515233)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33515233).
